### DISCUSSION TITLE
FixedDirection?

### DISCUSSION THREAD

**Discussion by colorindarkness** [Sixze/ALS-Refactored#69] [Q&A]
I'm wanting to implement a new RotationMode called FixedDirection, that will allow for Free Look (Character movement becomes independent of Camera rotation). 

Where would be best to implement this in AlsCharacter.cpp? 
Perhaps during AAlsCharacter::RefreshGroundedNotMovingAimingActorRotation?

**Comment by Sixze** [comment]
Adding a new rotation mode is not trivial, but as an alternative you can simply change the `AAlsCharacterExample::InputMove()` function to something like this:
```cpp
void AAlsCharacterExample::InputMove(const FInputActionValue& ActionValue)
{
	const auto Value{UAlsMath::ClampMagnitude012D(ActionValue.Get<FVector2D>())};

	AddMovementInput(FVector{Value.Y, Value.X, 0.0f});
}
```

**Reply by colorindarkness** [reply]
I'm trying to make it so that when FreeLooking is True, Pushing Forward will move forward according to the Mesh orientation when Free Look was activated, meanwhile the camera and Look At can move freely without changing the rotation of the player.

I got it working mostly. I had to also Disallow Turn/RotateInPlace if IsFreeLooking() in the Anim Instance. But I still cant disable it during First Person. Any ideas?

```
void AAlsCharacterExample::InputMove(const FInputActionValue& ActionValue)
{
	const auto Value{UAlsMath::ClampMagnitude012D(ActionValue.Get<FVector2D>())};

	FRotator CharRotation = GetMesh()->GetComponentRotation();

	const auto ForwardDirection{UAlsMath::AngleToDirectionXY(UE_REAL_TO_FLOAT(GetViewState().Rotation.Yaw))};
	const auto RightDirection{UAlsMath::PerpendicularCounterClockwiseXY(ForwardDirection)};

	const auto CharForwardDirection{ UAlsMath::AngleToDirectionXY(UE_REAL_TO_FLOAT(CharRotation.Pitch)) };
	const auto CharRightDirection{ UAlsMath::PerpendicularCounterClockwiseXY(CharForwardDirection) };

	if (IsFreeLooking())
	{
		AddMovementInput(CharForwardDirection * Value.Y + CharRightDirection * Value.X);
	}
	else
	{
		AddMovementInput(ForwardDirection * Value.Y + RightDirection * Value.X);
	}
}
```

**Reply by Sixze** [reply]
Check `AAlsCharacter::RefreshGroundedNotMovingAimingActorRotation()` and `AAlsCharacter::RefreshInAirAimingActorRotation` functions. They control how the character rotates when aiming or in first person mode.
